data/appRateLimiter.go

```diff
@@ -0,0 +1,249 @@
+
+	for i := 0; i < numBuckets; i++ {
+		r.buckets[i] = make(map[keyType]*appRateLimiterEntry)
+		r.lrus[i] = util.NewList[keyType]().AllocateFreeNodes(int(maxBucketSize))
+	}
+	return r
+}
+
+func (r *appRateLimiter) entry(b int, key keyType, curInt int64) (*appRateLimiterEntry, bool) {
+	r.mus[b].Lock()
+	defer r.mus[b].Unlock()
+
+	if len(r.buckets[b]) >= int(r.maxBucketSize) {
+		// evict the oldest entry
+		start := time.Now()
+		atomic.AddUint64(&r.evictions, 1)
+
+		el := r.lrus[b].Back()
+		delete(r.buckets[b], el.Value)
+		r.lrus[b].Remove(el)
+
+		atomic.AddUint64(&r.evictionTime, uint64(time.Since(start)))
+	}
+
+	entry, ok := r.buckets[b][key]
+	if ok {
+		el := entry.lruElement
+		r.lrus[b].MoveToFront(el)
+	} else {
+		el := r.lrus[b].PushFront(key)
+		entry = &appRateLimiterEntry{lruElement: el}
+		entry.cur.Store(1)
+		entry.interval.Store(curInt)
+		r.buckets[b][key] = entry
+	}
+	return entry, ok
+}
+
+// interval calculates the interval numeric representation based on the given time
+func (r *appRateLimiter) interval(now time.Time) int64 {
+	return now.UnixNano() / int64(r.serviceRateWindow)
+}
+
+// fraction calculates the fraction of the interval that is elapsed since the given time
+func (r *appRateLimiter) fraction(now time.Time) float64 {
+	return float64(now.UnixNano()%int64(r.serviceRateWindow)) / float64(r.serviceRateWindow)
+}
+
+// shouldDrop returns true if the given transaction group should be dropped based on the
+// on the rate for the applications in the group: the entire group is dropped if a single application
+// exceeds the rate.
+func (r *appRateLimiter) shouldDrop(txgroup []transactions.SignedTxn, origin []byte) bool {
+	return r.shouldDropInner(txgroup, origin, time.Now())
+}
+
+// shouldDropInner is the same as shouldDrop but accepts the current time as a parameter
+// in order to make it testable
+func (r *appRateLimiter) shouldDropInner(txgroup []transactions.SignedTxn, origin []byte, now time.Time) bool {
+	buckets, keys := txgroupToKeys(txgroup, origin, r.seed, r.salt, numBuckets)
+	if len(keys) == 0 {
+		return false
+	}
+	return r.shouldDropKeys(buckets, keys, now)
+}
+
+func (r *appRateLimiter) shouldDropKeys(buckets []int, keys []keyType, now time.Time) bool {
+	curInt := r.interval(now)
+
+	for i := range keys {
+		key := keys[i]
+		bucket := buckets[i]
+		entry, has := r.entry(bucket, key, curInt)
+		if !has {
+			// new entry, defaults are provided by entry() function
+			continue
+		}
+
+		interval := entry.interval.Load()
+		if interval != curInt {
+			var val uint64 = 0
+			if interval == curInt-1 {
+				// there are continuous intervals, use the previous value
+				val = entry.cur.Load()
+			}
+			entry.prev.Store(val)
+			entry.cur.Store(1)
+			entry.interval.Store(curInt)
+		} else {
+			entry.cur.Add(1)
+		}
+
+		curFraction := r.fraction(now)
+		rate := uint64(float64(entry.prev.Load())*(1-curFraction)) + entry.cur.Load()
+
+		if rate > r.serviceRatePerWindow {
+			return true
+		}
```

Comment:
The entry modifications aren't rolled back when the rate is exceeded.  So the rate is more of an attempted app rate than an admitted app rate.  A more nimble txgroup that might have been accepted had it gone first is now blocked by the effects of the prior dropped txgroups.

Reply:
nice catch, fixed